### PR TITLE
Micro batch loader for bert model

### PR DESCRIPTION
**Pull request by shanmugamr1992**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py [hunk 1]
```diff
@@ -359,21 +375,22 @@ def training_step(self, batch, batch_idx):
         if self.cfg.precision == 16:
             loss_scale = self.trainer.precision_plugin.scaler._scale
             if loss_scale is not None:
-                self.log('loss_scale', loss_scale)
+                self.log('loss_scale', loss_scale, batch_size=1)
```
Comment (yidong72) [inline nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:378]: what is this `batch_size=1` do here?

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py [hunk 2]
```diff
@@ -862,3 +866,41 @@ def on_load_checkpoint(self, checkpoint) -> None:
                 parallel_state.set_virtual_pipeline_model_parallel_rank(i)
                 self.model[i].module.load_state_dict(checkpoint[f'model{i}'], strict=True)
             parallel_state.set_virtual_pipeline_model_parallel_rank(0)
+
+    def on_train_batch_end(self, outputs, dataloader_iter: Any, batch_idx: int, unused: Optional[int] = 0) -> None:
+        super().on_train_batch_end(outputs, dataloader_iter, batch_idx)
+
+        # TODO: Replace with newer override for scheduler.step() instead of
```
Comment (yidong72) [inline nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:873]: why is this related to the new microbatch loader?

**Review by yidong72 — COMMENTED** [review]
Looks good. Just have two questions about this PR.

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py (hunk 1 above)
Comment (shanmugamr1992) [reply nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:378]: It throws an error otherwise. Pretty much pulled it from the gpt micro batch loader code . @erhoo82  might know better.

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py (hunk 2 above)
Comment (shanmugamr1992) [reply nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:873]: This as well pulled from gpt3 code . @erhoo82  might know better.

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py (hunk 1 above)
Comment (ericharper) [reply nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:378]: It's a hack to make PTL happy, we need to raise an issue with them.

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py (hunk 2 above)
Comment (ericharper) [reply nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:873]: We use that method so we need to override it with the datloader_iter signature

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!
